File: Olympiad/COCI/official/2008/contest2_solutions/prva.pas
program prva;

const MAX = 20;

var
   R, C : longint;
   rr, cc : longint;
   puzzle : array[1..MAX] of string;
   solution : string;
   word : string;

procedure append( x : char );
begin
   if x <> '#' then begin

      word := word + x;

   end else begin

      if (length(word) >= 2) and (word<solution) then solution := word;
      word := '';

   end;
end;

begin
   readln( R, C );

   for rr := 1 to R do readln( puzzle[rr] );

   solution := '~';

   for rr := 1 to R do begin
      for cc := 1 to C do
         append( puzzle[rr][cc] );
      append( '#' );
   end;

   for cc := 1 to C do begin
      for rr := 1 to R do
         append( puzzle[rr][cc] );
      append( '#' );
   end;

   writeln( solution );

end.

File: Olympiad/COCI/official/2008/contest2_solutions/prva.cpp
#include <cstdio>
#include <cstring>

#define MAX 20

int R, C;
char puzzle[MAX][MAX+1];

char solution[MAX+1] = "~";

int chars;
char word[MAX+1];

void append( char x ) {
   if( x != '#' ) {
      word[chars++] = x;
   } else {
      word[chars] = 0;
      if( chars >= 2 && strcmp( word, solution ) < 0 ) 
         strcpy( solution, word );
      chars = 0;
   } 
}

int main( void ) {
   scanf( "%d%d", &R, &C );
   for( int r = 0; r < R; ++r ) scanf( "%s", puzzle[r] );

   for( int r = 0; r < R; ++r ) {
      for( int c = 0; c < C; ++c ) 
         append( puzzle[r][c] );
      append( '#' );
   }

   for( int c = 0; c < C; ++c ) {
      for( int r = 0; r < R; ++r ) 
         append( puzzle[r][c] );
      append( '#' );
   }

   printf( "%s\n", solution );

   return 0;
}
